plugins/usbpro/OpenDeckDevice.cpp

```diff
@@ -0,0 +1,256 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * OpenDeckDevice.cpp
+ * An OpenDeck Device
+ * Copyright (C) 2021 Simon Newton, Igor Petrovic
+ *
+ * This device creates a single output port
+ */
+
+#include <iomanip>
+#include <ios>
+#include <iostream>
+#include <string>
+
+#include "common/rpc/RpcController.h"
+#include "ola/Callback.h"
+#include "ola/Constants.h"
+#include "ola/Logging.h"
+#include "plugins/usbpro/OpenDeckDevice.h"
+#include "plugins/usbpro/UsbProWidgetDetector.h"
+
+namespace ola {
+namespace plugin {
+namespace usbpro {
+
+using ola::plugin::usbpro::Request;
+using ola::plugin::usbpro::Reply;
+using ola::rpc::RpcController;
+using ola::utils::SplitUInt16;
+using std::string;
+
+/*
+ * Create a new device
+ * @param owner  the plugin that owns this device
+ * @param name  the device name
+ * @param dev_path  path to the pro widget
+ */
+OpenDeckDevice::OpenDeckDevice(ola::PluginAdaptor *plugin_adaptor,
+                               ola::AbstractPlugin *owner,
+                               const string &name,
+                               OpenDeckWidget *widget,
+                               OLA_UNUSED uint16_t esta_id,
+                               OLA_UNUSED uint16_t device_id,
+                               uint32_t serial,
+                               uint16_t firmware_version,
+                               unsigned int fps_limit):
+    UsbSerialDevice(owner, name, widget),
+    m_opendeck_widget(widget),
+    m_serial(),
+    m_got_parameters(false),
+    m_break_time(0),
+    m_mab_time(0),
+    m_rate(0) {
+  std::ostringstream str;
+  str << std::setfill('0');
+  uint8_t *ptr = reinterpret_cast<uint8_t*>(&serial);
+  for (int i = 0; i < UsbProWidgetInformation::SERIAL_LENGTH; i++) {
+    str << std::setfill('0') << std::setw(2) << std::hex
+        << static_cast<int>(ptr[i]);
+  }
```

Comment:
A comment here would be good to confirm the data format, or maybe even pull it out to it's own function so you can run a few tests against it. It looks like it's stored as hex digits and then turned to a string.

You may just be able to use our ToHex function:
http://docs.openlighting.org/ola/doc/latest/namespaceola_1_1strings.html#aedd1827ba98aeee25dbb9fdc8673d225

Reply:
Done